torch/lib/THD/base/data_channels/DataChannelGloo.cpp

```diff
@@ -0,0 +1,417 @@
+    case ::thpp::Type::FLOAT: func<float>(args); break;                       \
+    case ::thpp::Type::DOUBLE: func<double>(args); break;                     \
+    /* case ::thpp::Type::HALF: func<float>(args); break; */                  \
+    case ::thpp::Type::SHORT: func<short>(args); break;                       \
+    case ::thpp::Type::USHORT: func<unsigned short>(args); break;             \
+    case ::thpp::Type::INT: func<int>(args); break;                           \
+    case ::thpp::Type::UINT: func<unsigned int>(args); break;                 \
+    case ::thpp::Type::LONG: func<long>(args); break;                         \
+    case ::thpp::Type::ULONG: func<unsigned long>(args); break;               \
+    case ::thpp::Type::LONG_LONG: func<long long>(args); break;               \
+    case ::thpp::Type::ULONG_LONG: func<unsigned long long>(args); break;     \
+    default:                                                                  \
+      throw std::runtime_error("Invalid " + std::string(#func) + " function type"); \
+  }
+
+
+namespace thd {
+
+using ::gloo::rendezvous::PrefixStore;
+using ::gloo::rendezvous::Context;
+
+namespace {
+
+template<typename T>
+const ::gloo::ReductionFunction<T>* THDToGlooReduceOp(THDReduceOp op);
+
+} // anonymous namespace
+
+
+DataChannelGloo::RequestGloo::RequestGloo(QueueWorker::Request&& request)
+  : _request(std::move(request)) {
+}
+
+
+DataChannelGloo::RequestGloo::~RequestGloo() {}
+
+
+bool DataChannelGloo::RequestGloo::isCompleted() {
+  return _request.isCompleted();
+}
+
+
+void DataChannelGloo::RequestGloo::wait() {
+  _request.wait();
+}
+
+
+DataChannelGloo::DataChannelGloo()
+  : _rank(load_rank_env())
+{
+  _store = std::unique_ptr<Store>(new thd::Store());
+
+  ::gloo::transport::tcp::attr attr; // default options listen on all interfaces
+  _device = ::gloo::transport::tcp::CreateDevice(attr);
+  if (_rank == 0) {
+    _num_processes = load_world_size_env();
+    auto num_proc_str = std::to_string(_num_processes);
+    _store->set("world_size",
+            std::vector<char>(num_proc_str.begin(), num_proc_str.end()));
+  } else {
+    auto world_size = _store->get("world_size");
+    _num_processes = std::atoll(
+            std::string(world_size.begin(), world_size.end()).c_str());
+  }
+}
+
+
+DataChannelGloo::~DataChannelGloo() {
+}
+
+
+bool DataChannelGloo::init() {
+  std::vector<rank_type> ranks;
+  ranks.reserve(_num_processes);
+  for (rank_type rank = 0; rank < _num_processes; ++rank)
+    ranks.push_back(rank);
+
+  _groups.insert({
+    THDGroupWORLD,
+    DataChannel::Group(ranks, _num_processes - 1)
+  });
+
+  createContexts(THDGroupWORLD);
+  return true;
+}
+
+PrefixStore DataChannelGloo::getStore() {
+  static std::uint64_t id = 0; // TODO: that's not a good solution
+  return PrefixStore(std::to_string(id++), *_store);
+}
+
+void DataChannelGloo::createContexts(THDGroup group_id) {
+  // there is a separate context for every operation
```

Comment:
eh that's not really how I explained context caching to you... let's talk about that tomorrow